src/Mod/Cloud/App/AppCloud.cpp

```diff
@@ -0,0 +1,693 @@
+{
+             Cloud::CloudReader::FileList.push_back(new_entry);
+}
+
+void Cloud::CloudReader::checkXML(DOMNode* node) {
+     if (node) {
+           switch (node->getNodeType()) {
+           case DOMNode::ELEMENT_NODE:
+                checkElement(static_cast<DOMElement*>(node));
+                break;
+           case DOMNode::TEXT_NODE:
+                checkText(static_cast<DOMText*>(node));
+                break;
+           default:
+                break;
+           }
+           DOMNode* child = node->getFirstChild();
+           while (child) {
+                DOMNode* next = child->getNextSibling();
+                checkXML(child);
+                child = next;
+           }
+     }
+}
+
+
+
+struct data_buffer
+{
+        const char *ptr;
+        size_t remaining_size;
+};
+
+Cloud::CloudReader::~CloudReader()
+{
+}
+
+Cloud::CloudReader::CloudReader(const char* Url, const char* AccessKey, const char* SecretKey, const char* TcpPort, const char* Bucket) 
+{
+
+        struct timeval tv;
+        struct tm *tm;
+        char date_formatted[256];
+        char StringToSign[1024];
+        unsigned char *digest;
+        CURL *curl;
+        CURLcode res;
+
+        std::string s;
+
+
+        this->Url=Url;
+        this->AccessKey=AccessKey;
+        this->SecretKey=SecretKey;
+        this->TcpPort=TcpPort;
+        this->Bucket=Bucket;
+
+        // Amazon S3 and Swift require the timezone to be define to
+        // GMT. As to simplify the conversion this is performed through the TZ
+        // environment variable and a call to localtime as to convert output of gettimeofday
+        setenv("TZ","GMT",1);
+
+        // We must get the directory content
+
+
+        gettimeofday(&tv,NULL);
+        tm = localtime(&tv.tv_sec);
+        strftime(date_formatted,256,"%a, %d %b %Y %T %z", tm);
+
+        sprintf(StringToSign,"GET\n\napplication/xml\n%s\n/%s/", date_formatted, this->Bucket);
+
+        // We have to use HMAC encoding and SHA1
+        digest=HMAC(EVP_sha1(),this->SecretKey,strlen(this->SecretKey),
+                (const unsigned char *)&StringToSign,strlen(StringToSign),NULL,NULL);
+
+        // Let's build the Header and call to curl
+        curl_global_init(CURL_GLOBAL_ALL);
+        curl = curl_easy_init();
+        if ( curl )
+        {
+                struct curl_slist *chunk = NULL;
+                char header_data[1024];
+                // Let's build our own header
+                sprintf(header_data,"Host: %s:%s", strtok((char *)this->Url,"http://"),
+                                                   this->TcpPort);
+                chunk = curl_slist_append(chunk, header_data);
+                sprintf(header_data,"Date: %s", date_formatted);
+                chunk = curl_slist_append(chunk, header_data);
+                chunk = curl_slist_append(chunk, "Content-Type: application/xml");
+               std::string digest_str;

```

Comment:
Missing a space on the indentation here.

Reply:
I have updated the file, thanks to have pointed it out